test/lit/basic/polymorphic_stack.wast

```diff
@@ -219,8 +219,11 @@
   ;; CHECK-TEXT:      (func $untaken-break-should-have-value (type $0) (result i32)
   ;; CHECK-TEXT-NEXT:  (block $x (result i32)
   ;; CHECK-TEXT-NEXT:   (block
-  ;; CHECK-TEXT-NEXT:    (br_if $x
+  ;; CHECK-TEXT-NEXT:    (drop
   ;; CHECK-TEXT-NEXT:     (i32.const 0)
+  ;; CHECK-TEXT-NEXT:    )
+  ;; CHECK-TEXT-NEXT:    (br_if $x
+  ;; CHECK-TEXT-NEXT:     (unreachable)
   ;; CHECK-TEXT-NEXT:     (unreachable)
```

Comment:
This change may have implications for our testing. Before, we were able to make a particular argument to this `br_if` be unreachable, and test specifically the proper handling of that, while after, it looks like all arguments become unreachable. My concern is that unreachability of the first argument may prevent testing of handling of only the second argument being unreachable (which is possible after optimizations).

There is also a minor downside to having less precise roundtripping here when debugging.

How difficult would it be to make the parser roundtrip more accurately here?

Reply:
Consider these two instruction sequences:

```
i32.const 0
unreachable
i32.add

f32.const 0
unreachable
i32.add
```

They are both valid, and we currently parse both just fine:

```
(drop
 (i32.const 0)
)
(i32.add
 (unreachable)
 (unreachable)
)

(drop
 (f32.const 0)
)
(i32.add
 (unreachable)
 (unreachable)
)
```

If we tried to match existing behavior with the first sequence, we would end up with this for the second sequence:

```
(i32.add
 (f32.const 0)
 (unreachable)
)
```

Taking a look at our validator, it looks like this might be ok, actually? I don't know how consistent we are about letting children of unreachable nodes have any type.

(Also our validator will fail to reject `(i32.add (unreachable) (f32.const 0))`, but that's a separate issue.)

Reply:
Hmm, I think that would error - the parent still requires the non-unreachable children make sense.

I agree this is an issue for the unfolded form, but it seems that it has a downside for the folded form. And we use that folded form in order to test our IR, so being able to express less there is concerning.

Reply:
After talking offline, the plan is to try augmenting `IRBuilder` with some additional smarts around the expected types of things it pops. This will let the parser continue parsing instructions without having to worry about their nested relationships.

Today when an unreachable instruction is pushed, everything before it is automatically dropped. We will instead defer that automatic dropping behavior as far as possible. When popping the children for each newly-parsed instruction, we will attempt to pop existing instructions from the stack, even past an unreachable instruction. If we pop past an unreachable and encounter a mismatch between the expected and available type, we will undo the popping back to and including the unreachable, then fall back to the auto-drop behavior we have today. We will also auto-drop everything when finishing a scope.

I will implement this after getting fuzzing of text round-tripping working with the current implementation. I can probably extend subtype-exprs.h to pass a pointer to the child expression to `noteSubtype` and build the necessary functionality off of that. We could also reuse this utility in the validator.